Fake quantization support for half and double precision variables

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52590 Corrected linting and retested
* #52533 Trying to consolidate diffs
* #52532 Checking broken tests
* #52500 Finished GPU and backward for fake quantization support for half and double precision floats
* #52499 Fake Quantize for Double and Half precision variables
* **#52401 Fake quantization support for half and double precision variables**
* #52400 half precision fake quantization

Summary: cpu forward works for double and half precision, backwards and gpu incomplete

Test Plan:
python test/test_quantization.py TestFakeQuantization.test_forward_half_precision_fake_quantize
python test/test_quantization.py TestFakeQuantization.test_forward_double_precision_fake_quantize
python test_test_quantization.py TestFakeQuantization.test_backward_double_precision_fake_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26536118](https://our.internmc.facebook.com/intern/diff/D26536118)

HDCharles committed:
Fake quantization support for half and double precision variables

Summary: cpu forward works for double and half precision, backwards and gpu incomplete

Test Plan:
python test/test_quantization.py TestFakeQuantization.test_forward_half_precision_fake_quantize
python test/test_quantization.py TestFakeQuantization.test_forward_double_precision_fake_quantize
python test_test_quantization.py TestFakeQuantization.test_backward_double_precision_fake_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -2096,42 +2098,44 @@ void fake_quantize_learnable_tensor_grad_kernel_cpu(
-      } else {
-        *dZeroPointOutput = 0;
-        *dScaleOutput = (*dYInput) * (xfqi - (*XInput)) * inv_scale * grad_factor;
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.common_dtype(), "fake_quantize_backward_tensor_cachemask_kernel_types", [&] {
```
Technically learnable fake_quant is outside of the scope of the original issue.  Optional, but I'd recommend either not doing this at all (at this time), or doing it in a separate PR.  This will reduce the risk and speed up the testing.

HDCharles replied:
oh, alright, i'll undo this.

vkuzo commented on test/quantization/test_workflow_module.py:
```diff
@@ -822,13 +822,89 @@ def test_histogram_observer_against_reference(self, N, bins, dtype, qscheme, red
 
 class TestFakeQuantize(TestCase):
-    def test_half_precision_fake_quantize(self):
+    def test_forward_half_precision_fake_quantize(self):
```
nice.  For the tests, would it make sense to reuse the existing test code across dtypes?  Something like

```
def _test_forward_per_tensor_cachemask_impl(self, device):
  is_half_options = (True, False)
  torch_types = (torch.qint8, torch.quint8)
  test_cases = itertools.product(is_half_options, torch_types)
  for is_half, torch_type in test_cases:
    ... (test case code) ...
```

HDCharles replied:
yeah that would make sense

vkuzo reviewed (commented):
nice, the CPU part looks good!  Can we remove the learnable fake_quant from this PR, fix lint, update the test cases, and also add the GPU part?

As a follow up once we have the GPU part, it would also be great to actually test this with amp end to end.  That can be either in this PR or in a future one.

HDCharles committed:
Update on "Fake quantization support for half and double precision variables"


Summary: cpu forward works for double and half precision, backwards and gpu incomplete

Test Plan:
python test/test_quantization.py TestFakeQuantization.test_forward_half_precision_fake_quantize
python test/test_quantization.py TestFakeQuantization.test_forward_double_precision_fake_quantize
python test_test_quantization.py TestFakeQuantization.test_backward_double_precision_fake_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26536118](https://our.internmc.facebook.com/intern/diff/D26536118)

[ghstack-poisoned]

HDCharles commented:
> nice, the CPU part looks good! Can we remove the learnable fake_quant from this PR, fix lint, update the test cases, and also add the GPU part?
> 
> As a follow up once we have the GPU part, it would also be great to actually test this with amp end to end. That can be either in this PR or in a future one.

Yeah, I can do that probably tomorrow. I also think this whole diff web is a huge mess but I'm not 100% sure how to fix it, I thought the changes would be cleaned into a single diff but it looks like every single commit got its own github pull request.

HDCharles committed:
Update on "Fake quantization support for half and double precision variables"


Summary: cpu forward works for double and half precision, backwards and gpu incomplete

Test Plan:
python test/test_quantization.py TestFakeQuantization.test_forward_half_precision_fake_quantize
python test/test_quantization.py TestFakeQuantization.test_forward_double_precision_fake_quantize
python test_test_quantization.py TestFakeQuantization.test_backward_double_precision_fake_quantize

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26536118](https://our.internmc.facebook.com/intern/diff/D26536118)

[ghstack-poisoned]
